### PR TITLE
Refactor sdb_foreach and related APIs

### PR DESCRIPTION
**Pull request by p371jah**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [X] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [X] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

- `sdb_dump_dupnext()` was replaced with `sdb_dump_next()` which returns pointers to mmaped/buffered file (`sdb_const_get()` already does the same thing)
- `sdb_dump_next()` reads KV until start of hashes table is reached
- `sdb_foreach()` now passes ~key/value lengthes~ `SdbKv` to a callback
- `sdb_foreach_list*()` APIs were replaced with ~`sdb_get_kv_list*()` which return `RzList`~ `sdb_get_items*()` which return `RzPVector`  instead of `SdbList` (it is the first step of getting rid of `SdbList`)

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Review by XVilka — CHANGES_REQUESTED** [review]
Nice changes!

```
 Mismatched function annotation for sdbkv_free at <rizin/librz/util/sdb/src/sdb.h:172:13> : None (was RZ_NULLABLE at <rizin/librz/util/sdb/src/sdb.c:514:13>
Missing type comment at <rizin/librz/util/sdb/src/sdb.c:642:10> (token is not a comment)
Missing type comment at <rizin/librz/util/sdb/src/sdb.c:691:10> (token is not a comment)
Missing type comment at <rizin/librz/core/tui/classes.c:438:85> (token is not a comment)
```

**Comment by XVilka** [comment]
I wonder if it makes sense to return `RzPVector` in this case since we do not change the resulting `RzList` that we got from SDB

**Review by wargio — APPROVED** [review]

File: librz/util/sdb/src/sdb.c [hunk 1]
```diff
@@ -596,168 +600,186 @@ RZ_API int sdb_set(Sdb *s, const char *key, const char *val, ut32 cas) {
...
 
-RZ_API SdbList *sdb_foreach_list(Sdb *s, bool sorted) {
-	SdbList *list = ls_newf((SdbListFree)sdbkv_free);
-	sdb_foreach(s, sdb_foreach_list_cb, list);
+/**
+ * \brief Get list of all key-value pairs in DB
+ * \param s DB
+ * \param sorted If set to true sort the resulting list by key in lexicographic order
+ */
+RZ_API RZ_OWN RzPVector /*<SdbKv *>*/ *sdb_get_kv_list(RZ_NONNULL Sdb *s, bool sorted) {
```
Comment (XVilka) [inline librz/util/sdb/src/sdb.c]: Please rename the function to represent it's a vector, not a list

File: librz/util/sdb/src/sdb.c [hunk 2]
```diff
@@ -596,168 +600,186 @@ RZ_API int sdb_set(Sdb *s, const char *key, const char *val, ut32 cas) {
...
 
-static bool sdb_foreach_list_cb(void *user, const char *k, const char *v) {
-	SdbList *list = (SdbList *)user;
-	SdbKv *kv = RZ_NEW0(SdbKv);
-	/* seems like some k/v are constructed in the stack and cant be used after returning */
-	kv->base.key = strdup(k);
-	kv->base.value = strdup(v);
-	ls_append(list, kv);
-	return 1;
+static bool sdb_get_kv_list_cb(void *user, const char *k, ut32 klen, const char *v, ut32 vlen) {
```
Comment (XVilka) [inline librz/util/sdb/src/sdb.c]: Same, rename

File: librz/util/sdb/src/sdb.c [hunk 3]
```diff
@@ -596,168 +600,186 @@ RZ_API int sdb_set(Sdb *s, const char *key, const char *val, ut32 cas) {
...
-	struct foreach_list_filter_t u;
-	SdbList *list = ls_newf((SdbListFree)sdbkv_free);
+/**
+ * \brief Get list of key-value pairs in DB matching filter \p filter
+ * \param s DB
+ * \param filter Filter callback, a key-value pair is skipped if callback return false
+ * \param user User data which is passed to filter callback \p cb
+ * \param sorted If set to true sort the resulting list by key in lexicographic order
+ */
+RZ_API RZ_OWN RzPVector /*<SdbKv *>*/ *sdb_get_kv_list_filter(RZ_NONNULL Sdb *s, RZ_NONNULL SdbForeachCallback filter, RZ_NULLABLE void *user, bool sorted) {
```
Comment (XVilka) [inline librz/util/sdb/src/sdb.c]: Same, rename

File: librz/util/sdb/src/sdb.c [hunk 4]
```diff
@@ -596,168 +600,186 @@ RZ_API int sdb_set(Sdb *s, const char *key, const char *val, ut32 cas) {
...
-err:
-	sdbkv_free(kv);
-	return false;
 }
 
-RZ_API SdbList *sdb_foreach_list_filter_user(Sdb *s, SdbForeachCallback filter, bool sorted, void *user) {
-	struct foreach_list_filter_t u;
-	SdbList *list = ls_newf((SdbListFree)sdbkv_free);
+/**
+ * \brief Get list of key-value pairs in DB matching filter \p filter
```
Comment (XVilka) [inline librz/util/sdb/src/sdb.c]: Also update Doxygen description

File: librz/util/sdb/src/sdb.c [hunk 5]
```diff
@@ -596,168 +600,186 @@ RZ_API int sdb_set(Sdb *s, const char *key, const char *val, ut32 cas) {
...
-		kv->base.key = strdup(k);
-		kv->base.value = strdup(v);
-		ls_append(o->list, kv);
-		if (o->single) {
-			return false;
-		}
+	RzPVector /*<SdbKv *>*/ *list;
+};
+
+static bool kv_list_match_cb(void *user, const char *k, ut32 klen, const char *v, ut32 vlen) {
```
Comment (XVilka) [inline librz/util/sdb/src/sdb.c]: Same

File: librz/util/sdb/src/sdb.c [hunk 6]
```diff
@@ -596,168 +600,186 @@ RZ_API int sdb_set(Sdb *s, const char *key, const char *val, ut32 cas) {
...
-	sdb_foreach(s, sdb_foreach_match_cb, &o);
-	return list;
-}
+/**
+ * \brief Get list of key-value pairs in DB with keys matching expression \p expr
+ * \param s DB
+ * \param expr Expression string
+ * \param sorted If set to true sort the resulting list by key in lexicographic order
+ */
+RZ_API RZ_OWN RzPVector /*<SdbKv *>*/ *sdb_get_kv_list_match(RZ_NONNULL Sdb *s, RZ_NONNULL const char *expr, bool sorted) {
```
Comment (XVilka) [inline librz/util/sdb/src/sdb.c]: Same

File: librz/core/cmd/cmd_search.c [hunk 7]
```diff
@@ -2469,18 +2468,21 @@ static void rop_kuery(void *data, const char *input, PJ *pj) {
 	switch (*input) {
 	case 'q':
 		ls_foreach (db_rop->ns, it, ns) {
-			sdb_list = sdb_foreach_list(ns->sdb, false);
-			ls_foreach (sdb_list, sdb_iter, kv) {
+			RzPVector *sdb_list = sdb_get_kv_list(ns->sdb, false);
```
Comment (XVilka) [inline librz/core/cmd/cmd_search.c]: Rename the variable as well

File: librz/core/cbin.c [hunk 8]
```diff
@@ -91,16 +91,16 @@ RZ_API void rz_core_bin_export_info(RzCore *core, int mode) {
...
 	if (IS_MODE_RZCMD(mode)) {
 		rz_cons_printf("fs format\n");
 	} else if (IS_MODE_SET(mode)) {
 		rz_flag_space_push(core->flags, "format");
 	}
 	// iterate over all keys
-	SdbList *ls = sdb_foreach_list(db, false);
-	ls_foreach (ls, iter, kv) {
+	void **iter;
+	RzPVector *ls = sdb_get_kv_list(db, false);
```
Comment (XVilka) [inline librz/core/cbin.c]: Rename the variable

File: librz/arch/platform_target_index.c [hunk 9]
```diff
@@ -59,17 +59,17 @@ RZ_API void rz_platform_item_free(RzPlatformItem *item) {
...
 static bool sdb_load_platform_profile(RZ_NONNULL RzPlatformTargetIndex *t, RZ_NONNULL Sdb *sdb) {
 	rz_return_val_if_fail(t && sdb, false);
-	SdbKv *kv;
-	SdbListIter *iter;
-	SdbList *l = sdb_foreach_list(sdb, false);
-	char *argument_key, *comment, *name;
-	ls_foreach (l, iter, kv) {
+
+	void **iter;
+	RzPVector *l = sdb_get_kv_list(sdb, false);
```
Comment (XVilka) [inline librz/arch/platform_target_index.c]: Rename the variable

File: librz/arch/cc.c [hunk 10]
```diff
@@ -265,16 +265,18 @@ RZ_API const char *rz_analysis_cc_func(RzAnalysis *analysis, const char *func_na
...
 	RzList *ccl = rz_list_new();
-	SdbKv *kv;
-	SdbListIter *iter;
-	SdbList *l = sdb_foreach_list(analysis->sdb_cc, true);
-	ls_foreach (l, iter, kv) {
-		if (!strcmp(sdbkv_value(kv), "cc")) {
-			rz_list_append(ccl, strdup(sdbkv_key(kv)));
-		}
+	void **iter;
+	RzPVector *l = sdb_get_kv_list_filter(analysis->sdb_cc, filter_cc, NULL, true);
```
Comment (XVilka) [inline librz/arch/cc.c]: Variable

File: librz/arch/analysis.c [hunk 11]
```diff
@@ -712,41 +712,36 @@ RZ_API bool rz_analysis_noreturn_at(RzAnalysis *analysis, ut64 addr) {
...
 	// At first we read all noreturn functions from the Types DB
 	RzList *noretl = rz_type_noreturn_function_names(analysis->typedb);
 	// Then we propagate all noreturn functions that were inferred by
 	// the analysis process
-	SdbKv *kv;
-	SdbListIter *iter;
-	SdbList *l = sdb_foreach_list(analysis->sdb_noret, true);
-	ls_foreach (l, iter, kv) {
+	void **iter;
+	RzPVector *l = sdb_get_kv_list_filter(analysis->sdb_noret, filter_noreturn, NULL, false);
```
Comment (XVilka) [inline librz/arch/analysis.c]: Variable

**Review by XVilka — CHANGES_REQUESTED** [review]
LGTM but needs minor renaming

File: librz/util/sdb/src/sdb.c (hunk 1 above)
Comment (p371jah) [reply librz/util/sdb/src/sdb.c]: `sdb_get_kv_vector` sounds not good too. What do you think about `sdb_get_items`?

Comment (XVilka) [reply librz/util/sdb/src/sdb.c]: @pelijah yes, `sdb_get_items` sounds good.

**Review by XVilka — APPROVED** [review]

**Pull request merged**